### ISSUE TITLE
lexical lvalue under recursion

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2077]
Migrated from [rt.perl.org#3363](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3363) (status was 'resolved')

Searchable as RT3363$

**Comment by p5pRT** [comment]
### From @vanstyn

#### Created by @vanstyn



In a subroutine called recursively\, ref\-to\-substr of a lexical
appears to do the wrong thing when deeper than the first call&#8203;:

crypt% perl \-wl
sub a \{
&nbsp;  my $word = "$\_\[0\]";
&nbsp;  my $replace = "$\_\[1\]";
&nbsp;  print "word&#8203;: $word; replace&#8203;: $replace";
&nbsp;  my $ref = \\substr\($word\, 0\, 1\);
&nbsp;  $$ref = $replace;
&nbsp;  print "word now&#8203;: $word";
&nbsp;  a\($word\, "b"\) unless $replace eq "b";
\}
a\("$\_"\, "a"\) for "test";
\_\_END\_\_
word&#8203;: test; replace&#8203;: a
word now&#8203;: aest
word&#8203;: aest; replace&#8203;: b
word now&#8203;: aest
crypt% 

Same with 5\.005\_03 and 5\.004\_05\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.0:

Configured by hv at Thu Mar 23 11:41:17 GMT 2000.

Summary of my perl5 (revision 5.0 version 6 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.5-16, archname=i686-linux-64int
    uname='linux crypt.compulink.co.uk 2.2.5-16 #1 sun may 30 23:00:18 bst 1999 i686 unknown '
    config_args='-des -Doptimize=-g -O6 -Dprefix=/opt/perl-5.6.0.64bit -Duse64bitint'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=define use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-g -O6', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)
    cppflags='-DDEBUGGING -fno-strict-aliasing'
    ccflags ='-DDEBUGGING -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt
    libc=/lib/libc-2.1.1.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.0:
    /opt/perl-5.6.0.64bit/lib/5.6.0/i686-linux-64int
    /opt/perl-5.6.0.64bit/lib/5.6.0
    /opt/perl-5.6.0.64bit/lib/site_perl/5.6.0/i686-linux-64int
    /opt/perl-5.6.0.64bit/lib/site_perl/5.6.0
    /opt/perl-5.6.0.64bit/lib/site_perl
    .


Environment for perl v5.6.0:
    HOME=/home/hv
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/hv/bin:/usr/bin:/usr/local/bin:/bin:/usr/X11R6/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Jun 10\, Hugo said&#8203;:

>sub a \{
>  my $word = "$\_\[0\]";
>  my $replace = "$\_\[1\]";
>  print "word&#8203;: $word; replace&#8203;: $replace";
>  my $ref = \\substr\($word\, 0\, 1\);
&nbsp;   print "\\nref is of type "\, ref\($ref\)\, "\\n";
>  $$ref = $replace;
>  print "word now&#8203;: $word";
>  a\($word\, "b"\) unless $replace eq "b";
>\}
>a\("$\_"\, "a"\) for "test";

In 5\.005\_02\, this gives me&#8203;:

&nbsp;  word&#8203;: test; replace&#8203;: a
&nbsp;  ref is of type LVALUE
&nbsp;  word now&#8203;: aest
&nbsp;  word&#8203;: aest; replace&#8203;: b
&nbsp;  ref is of type SCALAR
&nbsp;  word now&#8203;: aest

In fact\, adding this into the subroutine&#8203;:

&nbsp;  for my $S \(0\.\.3\) \{
&nbsp;    for my $L \(0\.\.4\-$S\) \{
&nbsp;      print "substr\(\\$word\,$S\,$L\) = "\, ref\(\\substr\($word\,$S\,$L\)\)\, "\\n";
&nbsp;    \}
&nbsp;  \}

shows that\, in the first call to the function\, a reference to any of the
substrings are LVALUEs\, but in the second call\, they are SCALARs\.

Hmm\.  This is true even of NON\-lexicals \(removing the my\(\) declarations\)\.
I did that to $word\, $replace\, and $ref\.

Hmm\, if I change the call to a\($word\,"b"\) to&#8203;:

&nbsp;  @&#8203;\_ = \($word\,"b"\);
&nbsp;  goto &a unless $replace eq "b";

then everything is ok \-\- since there is no new caller level\, I would
assume\.  Hmm\.  This is odd&#8203;:

sub a \{
&nbsp;  my $word = shift;
&nbsp;  print $word\, "\\n";

&nbsp;  for my $S \(0\.\.3\) \{
&nbsp;    for my $L \(0\.\.4\-$S\) \{
&nbsp;      print "substr\(\\$word\,$S\,$L\) = "\, ref\(\\substr\($word\,$S\,$L\)\)\, "\\n";
&nbsp;    \}
&nbsp;  \}

&nbsp;  my $other = "TEST";
&nbsp;  a\($other\) if $word eq "test";
\}

a\($\_\, "a"\) for "test";


It tells me the same LVALUE/SCALAR pattern\.

**Comment by p5pRT** [comment]
### From @vanstyn

In \<200006101424\.PAA07026@&#8203;crypt\.compulink\.co\.uk>\, I wrote&#8203;:
:In a subroutine called recursively\, ref\-to\-substr of a lexical
:appears to do the wrong thing when deeper than the first call

This apparently occurs because&#8203;:
1\) pad variables are PADTMP when in a recursively entered scope
2\) pp\.c&#8203;:S\_refto says&#8203;:
&nbsp;  if \(SvPADTMP\(sv\)\)
&nbsp;    sv = newSVsv\(sv\);
3\) \.\. which calls sv\.c&#8203;:Perl\_sv\_setsv\(newcopy\, oldcopy\);
4\) \.\. which says&#8203;:
&nbsp;  if \(stype == SVt\_PVLV\)
&nbsp;    \(void\)SvUPGRADE\(dstr\, SVt\_PVNV\);
&nbsp;  \.\. and makes no other attempt to copy SVt\_PVLV\-ness to the result\.

I don't understand those two lines from sv\_setsv\, but I suspect the bug
is that S\_refto\(\) is calling newSVsv\(sv\)\, which does not preserve the
SVt\_PVLV nature of sv\.

Hugo


---
crypt% perl \-MDevel&#8203;::Peek \-wle 'sub a \{ my\($x\, $y\) = @&#8203;\_; my $z = \\substr\($x\, 0\, 1\); print Dump\($z\); $$z = $y; a\($x\, $y\) if $ARGV\[0\]\-\- \} a\("test"\, "a"\)' 1
SV = RV\(0x811c2e4\) at 0x8125e60
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(PADBUSY\,PADMY\,ROK\)
&nbsp;  RV = 0x8125ea8
&nbsp;  SV = PVLV\(0x8125fe8\) at 0x8125ea8
&nbsp;    REFCNT = 2
&nbsp;    FLAGS = \(PADMY\,GMG\,SMG\,pPOK\)
&nbsp;    IV = 0
&nbsp;    NV = 0
&nbsp;    PV = 0x811ae60 "t"\\0
&nbsp;    CUR = 1
&nbsp;    LEN = 2
&nbsp;    MAGIC = 0x8112700
&nbsp;      MG\_VIRTUAL = &PL\_vtbl\_substr
&nbsp;      MG\_TYPE = 'x'
&nbsp;    TYPE = x
&nbsp;    TARGOFF = 0
&nbsp;    TARGLEN = 1
&nbsp;    TARG = 0x8118c20
&nbsp;  SV = PV\(0x8106b94\) at 0x8118c20
&nbsp;    REFCNT = 2
&nbsp;    FLAGS = \(PADBUSY\,PADMY\,POK\,pPOK\)
&nbsp;    PV = 0x8106fb0 "test"\\0
&nbsp;    CUR = 4
&nbsp;    LEN = 5

SV = RV\(0x811c2e8\) at 0x8125f14
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(PADBUSY\,PADMY\,ROK\)
&nbsp;  RV = 0x8125f5c
&nbsp;  SV = PVNV\(0x8107724\) at 0x8125f5c
&nbsp;    REFCNT = 1
&nbsp;    FLAGS = \(POK\,pPOK\)
&nbsp;    IV = 0
&nbsp;    NV = 0
&nbsp;    PV = 0x810cd30 "a"\\0
&nbsp;    CUR = 1
&nbsp;    LEN = 2

crypt%

**Comment by p5pRT** [comment]
### From @gannett-ggreer

On Sat Jun 10 01&#8203;:17&#8203;:41 2000\, hv@&#8203;crypt\.compulink\.co\.uk wrote&#8203;:
>
> In a subroutine called recursively\, ref\-to\-substr of a lexical
> appears to do the wrong thing when deeper than the first call&#8203;:
> 
> crypt% perl \-wl
> sub a \{
>   my $word = "$\_\[0\]";
>   my $replace = "$\_\[1\]";
>   print "word&#8203;: $word; replace&#8203;: $replace";
>   my $ref = \\substr\($word\, 0\, 1\);
>   $$ref = $replace;
>   print "word now&#8203;: $word";
>   a\($word\, "b"\) unless $replace eq "b";
> \}
> a\("$\_"\, "a"\) for "test";
> \_\_END\_\_
> word&#8203;: test; replace&#8203;: a
> word now&#8203;: aest
> word&#8203;: aest; replace&#8203;: b
> word now&#8203;: aest
> crypt%
> 
> Same with 5\.005\_03 and 5\.004\_05\.

I created a patch and test based on Hugo's initialize suspicion\[1\]\,
which is to make S\_refto not create a new SV if the original is a
SVt\_PVLV so that the lvalue\-ness is preserved\.

http&#8203;://m\-l\.org/~perl/misc/0001\-Fix\-for\-RT\-3363\-lexical\-lvalue\-under\-recursion\.patch

http&#8203;://github\.com/greerga/perl/commit/6320261e8a2431f5cc8d0be6664028f85d842de1

http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=3363

1&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=3363\#txn\-10221

\-\- 
George Greer

**Comment by p5pRT** [comment]
### From @gannett-ggreer

<details><summary>0001-Fix-for-RT-3363-lexical-lvalue-under-recursion.patch</summary>

```diff
From 6320261e8a2431f5cc8d0be6664028f85d842de1 Mon Sep 17 00:00:00 2001
From: George Greer <perl@greerga.m-l.org>
Date: Sun, 18 Jul 2010 20:01:32 -0400
Subject: [PATCH] Fix for RT#3363: lexical lvalue under recursion

Add a patch and test based on Hugo's initialize suspicion as recorded in
the ticket, which is to make S_refto not create a new SV if the original
is a SVt_PVLV so that the lvalue-ness is preserved.
---
 pp.c            |    2 +-
 t/op/sub_lval.t |   19 ++++++++++++++++++-
 2 files changed, 19 insertions(+), 2 deletions(-)

diff --git a/pp.c b/pp.c
index a78c1cc..61a6097 100644
--- a/pp.c
+++ b/pp.c
@@ -546,7 +546,7 @@ S_refto(pTHX_ SV *sv)
 	SvTEMP_off(sv);
 	SvREFCNT_inc_void_NN(sv);
     }
-    else if (SvPADTMP(sv) && !IS_PADGV(sv))
+    else if (SvPADTMP(sv) && !IS_PADGV(sv) && SvTYPE(sv) != SVt_PVLV)
         sv = newSVsv(sv);
     else {
 	SvTEMP_off(sv);
diff --git a/t/op/sub_lval.t b/t/op/sub_lval.t
index c20ffac..222b88d 100644
--- a/t/op/sub_lval.t
+++ b/t/op/sub_lval.t
@@ -3,7 +3,7 @@ BEGIN {
     @INC = '../lib';
     require './test.pl';
 }
-plan tests=>71;
+plan tests=>72;
 
 sub a : lvalue { my $a = 34; ${\(bless \$a)} }  # Return a temporary
 sub b : lvalue { ${\shift} }
@@ -570,3 +570,20 @@ Execution of - aborted due to compilation errors.
     lval_decl = 5;
     is($x, 5, "subroutine declared with lvalue before definition retains lvalue. [perl #68758]");
 }
+
+{
+    my $result_3363;
+    sub a_3363 {
+        my ($word, $replace) = @_;
+        my $ref = \substr($word, 0, 1);
+        $$ref = $replace;
+        if ($replace eq "b") {
+            $result_3363 = $word;
+        } else {
+            a_3363($word, "b");
+        }
+    }
+    a_3363($_, "v") for "test";
+
+    is($result_3363, "best", "ref-to-substr retains lvalue-ness under recursion [perl #3363]");
+}
-- 
1.7.0.4


```
</details>

**Comment by p5pRT** [comment]
### From @xdg

On Sun\, Jul 18\, 2010 at 8&#8203;:15 PM\, George Greer via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> I created a patch and test based on Hugo's initialize suspicion\[1\]\,
> which is to make S\_refto not create a new SV if the original is a
> SVt\_PVLV so that the lvalue\-ness is preserved\.

Thank you\.

As a reminder to reviewers of this patch\, we're in a code\-freeze until
after v5\.13\.3 is released\.

\-\- David

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Jul 18 17&#8203;:15&#8203;:21 2010\, greerga wrote&#8203;:
> On Sat Jun 10 01&#8203;:17&#8203;:41 2000\, hv@&#8203;crypt\.compulink\.co\.uk wrote&#8203;:
> >
> > In a subroutine called recursively\, ref\-to\-substr of a lexical
> > appears to do the wrong thing when deeper than the first call&#8203;:
> >
> > crypt% perl \-wl
> > sub a \{
> >   my $word = "$\_\[0\]";
> >   my $replace = "$\_\[1\]";
> >   print "word&#8203;: $word; replace&#8203;: $replace";
> >   my $ref = \\substr\($word\, 0\, 1\);
> >   $$ref = $replace;
> >   print "word now&#8203;: $word";
> >   a\($word\, "b"\) unless $replace eq "b";
> > \}
> > a\("$\_"\, "a"\) for "test";
> > \_\_END\_\_
> > word&#8203;: test; replace&#8203;: a
> > word now&#8203;: aest
> > word&#8203;: aest; replace&#8203;: b
> > word now&#8203;: aest
> > crypt%
> >
> > Same with 5\.005\_03 and 5\.004\_05\.
> 
> I created a patch and test based on Hugo's initialize suspicion\[1\]\,
> which is to make S\_refto not create a new SV if the original is a
> SVt\_PVLV so that the lvalue\-ness is preserved\.

Thank you for the patch\. It seems this has been fixed since you
submitted it\, by Eric Brine’s lvalue fix\-ups \(2154eca7; bug \#67838\)\.
substr no longer returns TARG \(which is PADTMP\)\.

Without Eric’s patch\, it looks to me as though your patch would
introduce a bug\, in that
&nbsp;  for\(1\,2\) \{ push @&#8203;refs\, \\substr $\_\,0\,1 \}
would cause the \\ to return the same sv on both iterations\. I’ve applied
a test for refgen\+PADTMP as e47aeb2\.

I’ve applied your tests as 645d8892bdf2a7e5afe985dfd321b2d332930f62\.

This code path is still buggy though\. See this bug that I’ve just
submitted as \#78194&#8203;:

$ perl  \-le' for\( "$y" \)\{print \\$\_\.\\$\_\}'
SCALAR\(0x8039f0\)SCALAR\(0x8038d0\)

\(The same address should be printed twice\.\)

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
